### DISCUSSION TITLE
How do I trigger a command after waiting a certain period of time (non blocking wait)?

### DISCUSSION THREAD

**Discussion by dheadrick1618** [nasa/fprime#2038] [Q&A] [answered]
I have had difficulty finding any reference or example of how to trigger a command after a certain interval of time. I know of rate groups but it doesn't seem like the right tool for the job here. I am creating a 'burn wire' component that will toggle a GPIO high for about 20 seconds and then toggle it low again, but this command will likely only run once, (or a few times at most). How should I go about this in a 'non-blocking' manner using Fprime (rather than timers with C++) ?

**Comment by SterlingPeet** [comment, marked as answer]
First off, you probably need an internal interface to allow your burnwire call to be triggered internally to the mission logic housed in a specific component, or externally by a ground command.  This is a common pattern, so its good to know how to do it.

The GT-1 mission timed their burnwires by counting rate group calls into the burnwire controller component.  The number of calls per second was known as a mission design parameter, and the number of calls to expire the counter was set as a parameter.  This allowed the flight software team to hand the responsibility of the burnwire timing to the testing and operations teams, who just calculated the length of time they wanted and set the parameter appropriately.

PS:  It worked beautifully, GT-1 deployed all of its solar panels on the first try, and was power positive until it de-orbited last June.

**Reply by dheadrick1618** [reply]
It will most likely be configured to be triggered internally, a set amount of time after deployment. I am currently implementing this design where an 'activate burn wire' command sets the burnwire state to on, and sets a count variable to zero. A run_handler function's input port is attached to a 1Hz rate group. When the state is set to it on will toggle a GPIO high if the count is zero, and for each successive call increment a counter until the counter hits a certain value, where the gpio will be toggled low, the counter reset, and the state set to off. The issue I have with this implementation (likely a result of my lack of understanding of Fprime), is **how I can 'disable' this run_handler input port once the burnwire sequence is complete**. It doesn't make sense to continue to run through the run_handler fxn (which at this point will do nothing because the state will remain off), every second for the remainder of the mission after the burnwire sequence is completed. Please let me know what you think I should do in this situation, or If I am potentially approaching this incorrectly.

**Reply by SterlingPeet** [reply]
Well, you definitely don't want that GPIO to get toggled to the burn state by an errant SEU with no method to turn it off again.  That would likely be an annoyingly high power drain on your system.  Writing it to the off state every second during the lifetime of the mission would mitigate that possibility.  By not providing a ground command to fire the burnwires, you preclude the ability of the operations team to attempt recovery from an off-nominal situation where the burnwires didn't operate such that the deployables actuated as expected.

GT-1 used a state machine that that driven by the rate group, and only when it was in the initial startup state did it make an attempt to operate the burnwires.  The nominal operations state just didn't make those port calls into the burnwire component.  GT-1's startup state was also a configurable parameter, so the state machine could be set to start up into the nominal mode instead of going through the initial deployment sequences, at the discretion of the operations team.  I don't recall off-hand if the burnwire GPIO was being actively set for the the whole mission, but I generally advocate for belt-and-suspenders approach because you can't just go push the reset button.

State machine mission operations components are a fairly common pattern, but they are not well described in the documentation and there aren't any public examples that I am aware of.  Because they are so specific to the mission, this type of component is not a good candidate for reuse, hence the lack of examples.  @LeStarch this would probably be a good how-to deep dive topic, distinct from getting started tutorials.

**Reply by dheadrick1618** [reply]
Great points. Thanks for clarifying this for me.
